### PR TITLE
Resumable transformations

### PR DESCRIPTION
**Pull request by MichaelBelousov**
## Changes

- add `saveStateToFile` and `resumeTransformation` functions which save the internal state of a transformation to a sqlite database and allow you to create a new transformater instance from such a file. 
- add a disabled, locally-tested, "gauntlet" which simulates crashing in transformer APIs and especially native db methods, measuring the speed with which a resumed transformation is finished. This will be added to future transformer performance tests as enabled
- some fixes to false positives in `assertIdentityTransform`, as well as a way to use it with any id remapping
  - allow deepEqualWithFPTolerance to consider undefined fields as equal to non-existent fields. Useful for some JSON property equivalence testing 
- [Corresponding native PR](https://dev.azure.com/bentleycs/iModelTechnologies/_git/imodel02/pullrequest/245611)

## Motivation

Some agent orchestration frameworks like kubernetes will downscale by killing clusters/machines. This decreases costs of running many agents. It will do so by sending a sigterm for the process to handle, and then killing after a configurable grace period, the default being 30s (although in our actual env we have set this much higher).
This new feature allows long running transformations to be saved and resumed on other available agents to handle this case, since downscaling killing the process has a significant likelihood in long running jobs and otherwise the job would need to be restarted from scratch wasting a significant amount of computational resources.

### CODE REVIEW

**Review by ColinKerr — APPROVED** [review]

**Comment by MichaelBelousov** [comment]
When someone has a chance, could I get a review from @iTwin/itwinjs-core-admins before merging? (and same for the corresponding native PR, ask me for a link)

File: core/transformer/src/IModelTransformer.ts
Comment (pmconne) [inline core/transformer/src/IModelTransformer.ts:1198]: How likely is it that custom transformers will indeed need to override this? It looks awkward to do so since the base implementation is very self-contained (create db, write to it, close it, return nothing).
Also is there any risk of one custom transformer finding a state db leftover from a different custom transformer, attempting to resume, and being only partially constructed?

File: core/transformer/src/test/standalone/IModelTransformerResumption.test.ts
Comment (pmconne) [inline core/transformer/src/test/standalone/IModelTransformerResumption.test.ts:97]: Tests and randomness are generally not a good combo...

Comment (MichaelBelousov) [reply core/transformer/src/test/standalone/IModelTransformerResumption.test.ts:97]: the randomness is not used by default. It is behind a skipped/disabled test which you can activate to get some statistics of the speed of resumed transformations and whether contrived thrown errors throughout the stack generate a different result when resumed from a non-crashing transformation.

I enabled the test locally and ran several thousand transformations to test if resuming after various different crash points caused different results. I probably should put it somewhere, maybe in a work item, but I intend to run this in the transformer performance tests when those are setup. Just not as part of the unit test suite.

Comment (pmconne) [reply core/transformer/src/test/standalone/IModelTransformerResumption.test.ts:97]: Thanks for clarifying. Did it resume correctly when you tested locally?

File: core/transformer/src/IModelTransformer.ts
Comment (MichaelBelousov) [reply core/transformer/src/IModelTransformer.ts:1198]: It is certainly significantly likely. They don't have to store it in the database itself, but I think because I'm anticipating people wanting to do that to have one self contained resumable dump, I am going to add a new protected method with access to the `SqliteDb` object, which is going to be the override point. Please let me know if you have any concerns with that approach.

As for attempting to resume a dump from the wrong class, I will add an additional item to the json state to id the transformer  class used to make the dump. That can be used to sanity check. Thank you for pointing it out.

File: core/transformer/src/test/standalone/IModelTransformerResumption.test.ts
Comment (MichaelBelousov) [reply core/transformer/src/test/standalone/IModelTransformerResumption.test.ts:97]: Yes. The test transforms without the crashes enabled, and with, checks that crashes did occur, and then confirms that iModels that were crashed and resumed as well as iModels that ran to completion without classes have the same content, using the  `assertIdentityTransformation` function which asserts equality of as much as it can about entities, relationships, properties, between two imodels using some remapping scheme such as that in the completed transformer

File: core/transformer/src/IModelTransformer.ts
Comment (MichaelBelousov) [reply core/transformer/src/IModelTransformer.ts]: I have added tests demonstrating failing on attempting to resume from a different importer, exporter, or transformer class, based on the name of the class/constructor function which is now saved to the dump
.
I also added a few new `additionalData` functions for making it easier for custom transformers to store custom state without accessing the sqlitedb, and made it easier for them to access the sqlitedb in case they need to store it there. This has a side effect of cleaning up the code a tiny bit. There is also a test demonstrating that resuming does in fact pick up additional data across all the added new functions.

File: core/transformer/src/IModelTransformer.ts
Comment (pmconne) [inline core/transformer/src/IModelTransformer.ts:1236]: I can think of some ways in which `constructor.name` won't suffice (import IModelTransformer using an alias and define a subclass called IModelTransformer; instantiate IModelTransformer then add properties and replace methods) but people would have to go out of their way to break it.

**Review by pmconne — APPROVED** [review]

**Comment by MichaelBelousov** [comment]
oops merged master instead of imodel02, reverting

**Pull request merged**
